### PR TITLE
chore(ci): CTB Static Analysis Job

### PR DESCRIPTION
**Pull request by refcell**
**Description**

Modifies the existing `contracts-bedrock-slither` circleci job that runs the `packages/contracts-bedrock` slither check package target.

`slither:check` runs the `slither.sh` bash script, gracefully generating a slither report, and performing a `git diff` to compare it to the tracked `slither-report.json` file.

Fixes https://github.com/ethereum-optimism/security-pod/issues/67

## Summary by CodeRabbit

- **New Features**
  - Added a new static analysis job to the CI workflow for improved code quality and security checks.

- **Improvements**
  - Enhanced the static analysis process to generate more informative output for developers.

- **Chores**
  - Optimized the build process by removing unnecessary cache directory removal commands from the Dockerfile.

- **Documentation**
  - Included instructions for installing the static analysis tool when not found in the environment.

### CODE REVIEW

**Comment by refcell** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #8192** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/8192?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/ethereum-optimism/optimism/8192?utm_source=stack-comment).

File: packages/contracts-bedrock/scripts/slither.sh [hunk 1]
```diff
@@ -2,12 +2,31 @@
...
+dir=$(dirname "$0")
+
 # See slither.config.json for slither settings
 if [[ -z "$TRIAGE_MODE" ]]; then
-  echo "Building contracts"
-  forge build --build-info --force
-  echo "Running slither"
-  slither --ignore-compile .
+  echo "Running slither in normal mode"
+  slither . &> /dev/null || true
```
[the remark this answers is not in the document]
Comment (refcell) [reply packages/contracts-bedrock/scripts/slither.sh]: JSON Report contains all useful output. If anything fails, untracked changes to the `slither-report.json` file will cause ci to error.

File: packages/contracts-bedrock/scripts/slither.sh [hunk 2]
```diff
@@ -2,12 +2,31 @@
 
 rm -rf artifacts forge-artifacts
 
+set -e
+
+if ! command -v slither &> /dev/null
+then
+    echo "slither could not be found. Please install slither by running:"
+    echo "pip3 install slither-analyzer"
+    exit
```
[the remark this answers is not in the document]
Comment (refcell) [reply packages/contracts-bedrock/scripts/slither.sh]: Good rabbit

File: .circleci/config.yml [hunk 3]
```diff
@@ -378,6 +378,24 @@ jobs:
           working_directory: packages/contracts-bedrock
           no_output_timeout: 15m
 
+  contracts-bedrock-static-analysis:
+    docker:
+      - image: <<pipeline.parameters.ci_builder_image>>
+    resource_class: xlarge
```
Comment (clabby) [inline .circleci/config.yml]: Does the slither job need xlarge?

File: packages/contracts-bedrock/slither-report.json [hunk 4]
```diff
@@ -0,0 +1,1177 @@
+{
```
Comment (clabby) [inline packages/contracts-bedrock/slither-report.json]: Can we format this as file as an array of these objects? Right now these findings can't be parsed

**Review by clabby — CHANGES_REQUESTED** [review]
Looking good! Only thing looks to be serialization of the findings in the report

File: packages/contracts-bedrock/scripts/slither.sh [hunk 5]
```diff
@@ -2,12 +2,31 @@
...
+fi
+
+if [ -e "slither-report.json" ]; then
+    rm "slither-report.json"
+    echo "slither-report.json deleted."
+fi
+
+dir=$(dirname "$0")
```
Comment (tynes) [inline packages/contracts-bedrock/scripts/slither.sh]: will this work no matter what the homedir is when calling the script?

File: packages/contracts-bedrock/scripts/slither.sh [hunk 6]
```diff
@@ -2,12 +2,31 @@
...
+
 # See slither.config.json for slither settings
 if [[ -z "$TRIAGE_MODE" ]]; then
-  echo "Building contracts"
-  forge build --build-info --force
-  echo "Running slither"
-  slither --ignore-compile .
+  echo "Running slither in normal mode"
+  slither . &> /dev/null || true
```
Comment (tynes) [inline packages/contracts-bedrock/scripts/slither.sh]: why or true?

File: .circleci/config.yml [hunk 7]
```diff
@@ -378,6 +378,24 @@ jobs:
...
+    steps:
+      - checkout
+      - check-changed:
+          patterns: contracts-bedrock,op-node
+      - run:
+          name: print slither version
```
Comment (tynes) [inline .circleci/config.yml]: What do you think about removing this step and just running the version command in the bash script? Makes it more reproducible locally

File: packages/contracts-bedrock/scripts/slither.sh (hunk 6 above)
Comment (refcell) [reply packages/contracts-bedrock/scripts/slither.sh]: There's currently around 60 report items. If one is generated this command exits unsuccessfully. This pr has not triaged all these errors. If we want to triage I highly suggest stacking ontop and giving this task to the security team.

File: .circleci/config.yml (hunk 3 above)
Comment (refcell) [reply .circleci/config.yml]: What would you suggest it be?

File: ops/docker/ci-builder/Dockerfile [hunk 8]
```diff
@@ -106,7 +106,6 @@ RUN /bin/sh -c set -eux; \
   rm -rf /root/.cache/pip; \
   rm -rf /root/.cache/npm;
```
[the remark this answers is not in the document]
Comment (refcell) [reply ops/docker/ci-builder/Dockerfile]: wut

File: packages/contracts-bedrock/slither.config.json [hunk 9]
```diff
@@ -1,15 +1,24 @@
...
+    "exclude_informational": true,
+    "exclude_optimization": true,
+    "exclude_low": true,
+    "json": "slither-report.json",
+    "exclude_medium": false,
+    "exclude_high": false,
+    "solc_disable_warnings": false,
+    "disable_color": false,
+    "exclude_dependencies": true,
+    "filter_paths": "(lib/|test/|scripts/|src/vender|src/cannon/MIPS.sol)",
```
Comment (0xalpharush) [inline packages/contracts-bedrock/slither.config.json]:
```suggestion
    "filter_paths": "(lib/|src/vender|src/cannon/MIPS.sol)",
```
Slither will use Foundry's skip flag to skip tests and scripts by default as of 0.10.0

File: packages/contracts-bedrock/slither.config.json [hunk 10]
```diff
@@ -1,15 +1,24 @@
...
+    "foundry_out_directory": "artifacts",
+    "solc_remaps": [
+        "ds-test/=lib/ds-test/src/",
+        "forge-std/=lib/forge-std/src/",
+        "solmate/=lib/solmate/src/",
+        "@openzeppelin/=lib/openzeppelin-contracts/",
+        "safe-contracts/=lib/safe-tools/lib/safe-contracts/contracts/",
+        "safe-tools/=lib/safe-tools/src/",
+        "properties/=lib/properties/contracts/"
+    ]
```
Comment (0xalpharush) [inline packages/contracts-bedrock/slither.config.json]:
```suggestion
```
Remappings are entirely handled by foundry and not required

**Comment by 0xalpharush** [comment]
I would highly recommend using the sarif output with [github's code scanning](https://github.com/crytic/slither-action#github-code-scanning-integration). I believe it will remember results that are ignored, so you can keep detectors enabled and allow PR reviewers to see the result / determine whether to accept. It will look/work very similar to Semgrep

File: packages/contracts-bedrock/slither.config.json (hunk 9 above)
Comment (mds1) [reply packages/contracts-bedrock/slither.config.json]: Also typo here in the path name
```suggestion
    "filter_paths": "(lib/|src/vendor|src/cannon/MIPS.sol)",
```

**Comment by refcell** [comment]
> I would highly recommend using the sarif output with [github's code scanning](https://github.com/crytic/slither-action#github-code-scanning-integration). I believe it will remember results that are ignored, so you can keep detectors enabled and allow PR reviewers to see the result / determine whether to accept. It will look/work very similar to Semgrep

Thanks for your review! That would be cool, but we use circleci heavily and it definitely does not make sense to me to mix github actions and circleci jobs. Ideally, our infra team migrates all circleci jobs to github actions with custom or large hosted runners, but alas, much do little time

**Comment by refcell** [comment]
> Review Status

Need... approval... from rabbit

**Review by clabby — APPROVED** [review]

File: packages/contracts-bedrock/package.json [hunk 11]
```diff
@@ -27,13 +27,14 @@
...
+    "slither:check": "./scripts/slither.sh && git diff --exit-code",
+    "slither:triage": "TRIAGE_MODE=1 ./scripts/slither.sh",
     "semver-lock": "forge script scripts/SemverLock.s.sol",
     "validate-deploy-configs": "./scripts/check-deploy-configs.sh",
     "validate-spacers:no-build": "npx tsx scripts/validate-spacers.ts",
     "validate-spacers": "pnpm build && pnpm validate-spacers:no-build",
-    "slither": "./scripts/slither.sh",
-    "slither:triage": "TRIAGE_MODE=1 ./scripts/slither.sh",
-    "clean": "rm -rf ./artifacts ./forge-artifacts ./cache ./tsconfig.tsbuildinfo ./tsconfig.build.tsbuildinfo ./scripts/go-ffi/go-ffi ./.testdata ./deployments/hardhat/*",
+    "clean": "rm -rf ./artifacts ./forge-artifacts ./cache ./tsconfig.tsbuildinfo ./tsconfig.build.tsbuildinfo ./scripts/go-ffi/go-ffi ./.testdata",
```
Comment (trianglesphere) [inline packages/contracts-bedrock/package.json]: why do we not remove the hardhat deployments on the clean step? This seems like an important part of resetting to a fresh environement.

**Comment by refcell** [comment]
@tynes 👉 👈

**Review by tynes — APPROVED** [review]

**Pull request merged**